subprojects/core/src/main/java/org/gradle/api/internal/provider/GradlePropertiesCredentialsProviderFactory.java

```
@@ -0,0 +1,125 @@
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.Task;
+import org.gradle.api.credentials.PasswordCredentials;
+import org.gradle.api.execution.TaskExecutionGraph;
+import org.gradle.api.internal.properties.GradleProperties;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.TaskInputs;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.credentials.DefaultPasswordCredentials;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GradlePropertiesCredentialsProviderFactory implements CredentialsProviderFactory {
+
+    private final GradleProperties gradleProperties;
+    private final TaskExecutionGraph taskExecutionGraph;
+
+    private boolean credentialsBeingEvaluated = false;
+
+    private final Map<String, PasswordCredentialsProvider> passwordProviders = new HashMap<>();
+
+    public GradlePropertiesCredentialsProviderFactory(GradleProperties gradleProperties, TaskExecutionGraph taskExecutionGraph) {
+        this.gradleProperties = gradleProperties;
+        this.taskExecutionGraph = taskExecutionGraph;
+    }
+
+    @Override
+    public Provider<PasswordCredentials> usernameAndPassword(String identity) {
+        validateIdentity(identity);
+        registerCredentialsEvaluator();
+        return passwordProviders.computeIfAbsent(identity, PasswordCredentialsProvider::new);
+    }
+
+    private void registerCredentialsEvaluator() {
+        if (!credentialsBeingEvaluated) {
+            taskExecutionGraph.whenReady(graph -> {
+                graph.getAllTasks().stream().map(Task::getInputs).forEach(TaskInputs::getProperties);
+                passwordProviders.values().stream().filter(p -> p.valueRequested).forEach(AbstractMinimalProvider::get);
+            });
```

This is one way to validate credential presence before execution only when it is known that they will be needed and fail fast in such cases.
One (and probably major) downside is that it requires to traverse the task graph and evaluate all the declared inputs.